### ISSUE TITLE
Making .so on x86 w/ libz-ng.a

### ISSUE THREAD

**Issue by rjzak** [zlib-ng/zlib-ng#1759]
I'm trying to compile a Python module which links against the static library for Zlib-ng, and I was getting a linker error:

`relocation R_X86_PC32 against symbol zng_z_errmsg can not be used when making a shared object; recompile with -fPIC`

This is on a Debian 12 system on x86_64 with gcc. I was able to fix this by adding `-fPIC` to the CFLAGS variable in the Makefile after running the configure script. Maybe this could be added to the project?

Interestingly enough, I was able to compile the same code on Power9, also on Debian 12 with gcc without any issues.

**Comment by mtl1979** [comment]
This has been discussed numerous times before... Mixing static and shared libraries isn't supported as linker won't know how to resolve if two compilation units define or use same symbol. Some platforms might enable PIC or PIE by default, but that doesn't eliminate the root cause of why we don't support manually specifying `-fPIC`.

**Comment by rjzak** [comment]
Fair enough. In general, how would other projects handle this (shared and static), or how should it be done?

**Comment by mtl1979** [comment]
> Fair enough. In general, how would other projects handle this (shared and static), or how should it be done?

The main binary should be dynamically linked against zlib-ng and the symbols should be re-exported for any modules. This is the best way unless modules of that program can request other modules to be loaded on-demand.

**Comment by rjzak** [comment]
Sure, but the configure script has the option `--static`. Seems like that doesn't go well with not having `-fPIC`, since the linker will complain when a project then uses the resulting .a file. Maybe I'm not understanding correctly, I'm just trying to reduce run-time dependencies for easier distribution of my app.

**Comment by mtl1979** [comment]
> Sure, but the configure script has the option `--static`. Seems like that doesn't go well with not having `-fPIC`, since the linker will complain when a project then uses the resulting .a file. Maybe I'm not understanding correctly, I'm just trying to reduce run-time dependencies for easier distribution of my app.

The static flag only works when linking against an executable, not a shared library, that is limitation of the linker, not zlib-ng.

**Comment by rjzak** [comment]
What I meant was the project seems to support a static version of zlib-ng, and it seems that `-fPIC` is required for that. So I'm saying `-fPIC` seems to be required.

**Comment by nmoinvaz** [comment]
Documented here on the Wiki:
https://github.com/zlib-ng/zlib-ng/wiki/Build-tips#building-a-position-independent-static-library
https://github.com/zlib-ng/zlib-ng/wiki/Common-build-problems#relocation-error-linking-libz-nga-into-a-shared-object

Regarding Power compilation working, Claude says: 
> ppc64le's TOC-based ABI avoids the offending relocation, so the same build links there without PIC
